cmd/kyverno/main.go

```diff
@@ -157,14 +159,31 @@ func main() {
 	kyvernoV1beta1 := kyvernoInformer.Kyverno().V1beta1()
 	kyvernoV1alpha2 := kyvernoInformer.Kyverno().V1alpha2()
 
+	var registryOptions []registryclient.Option
+
 	// load image registry secrets
 	secrets := strings.Split(imagePullSecrets, ",")
 	if imagePullSecrets != "" && len(secrets) > 0 {
 		setupLog.Info("initializing registry credentials", "secrets", secrets)
-		if err := registryclient.Initialize(kubeClient, config.KyvernoNamespace(), "", secrets); err != nil {
-			setupLog.Error(err, "failed to initialize image pull secrets")
-			os.Exit(1)
-		}
+		registryOptions = append(
+			registryOptions,
+			registryclient.WithKeychainPullSecrets(kubeClient, config.KyvernoNamespace(), "", secrets),
+		)
+	}
+
+	if allowInsecureRegistry {
+		setupLog.Info("initializing registry with allowing insecure connections to registries")
+		registryOptions = append(
+			registryOptions,
+			registryclient.WithAllowInsecureRegistry(),
+		)
+	}
+
+	// initialize default registry client with our settings
+	registryclient.DefaultClient, err = registryclient.InitClient(registryOptions...)
```

Comment:
Can we keep the `DefaultClient` private to the `registryclient` package ?
I would prefer to keep this hidden from outside if possible, we don't want to assign the default client from somewhere else right ?

Reply:
Yeah, I can do that if that is required.
But, honestly, I don't see a big deal with that. This pattern is widely used in the standard library and in other libs as well - https://github.com/golang/go/blob/master/src/net/http/client.go#L109
https://github.com/golang/go/blob/master/src/encoding/base64/base64.go#L99
https://github.com/google/go-containerregistry/blob/main/pkg/authn/keychain.go#L55

But at the same time, I would rather remove this static DefaultClient and pass this object as a parameter variable to allow dependency injection. But unfortunately, the current code uses a lot of static functions/objects and requires huge refactoring to make it happen.

Reply:
Yeah, dependency injection would be great but i agree it's a big change.
How about keeping it private to the package and exposing a couple of static funcs ?
For example, the user should not be concerned with refreshing creds from secrets (this can be done in the package without requiring the user to think about calling the func himself).

Reply:
I understand the concern about modifying the default client, and I can make it private and provide a static function to work with this object. 

But I don't understand why we need to provide other static functions, it would make code less testable, especially parts with communication to external service (in our case communication to the k8s api to pull secrets, and requests to fetch digest from registry).

As for hidding logic of pulling secret from user of this package, yeah it would be nice. But in case if we make it dumb and simple (by fetching secrets every time when we requesting keychain) - it can be not optimal solution.
I don't see the whole picture of how everything works here, but I imagine the next use case. We provided 2 pull secrets, and we trying to validate one pod. This pod has 3 containers with different images. Then during this validation we would fetch those 2 secrets 3 times?
I would use other approach here, I think we can "subscribe" to the secret updates and handle them inside of this package. Or we can update pull secrets every N minutes.

Reply:
We can try to be smart indeed, but I feel very uncomfortable with this kind of code:
```
	if err := registryclient.DefaultClient.RefreshKeychainPullSecrets(); err != nil {
		return fmt.Errorf("unable to load image registry credentials, %w", err)
	}
	desc, err := registryclient.DefaultClient.FetchImageDescriptor(ref)
	if err != nil {
		return err
	}
       ....
```
In the end, I feel like it's not that different than doing `registryclient.FetchImageDescriptor(ref)` and letting the implementation refresh keychain secrets if necessary.

I mean, do we really expect the caller to refresh keychain secrets before calling other funcs on the client ?
I would rather not let this in the hands of the caller, really.

Reply:
Don't get me wrong, I think that what you did is already a nice improvement over what we had before, and thanks for that.

I was just trying to get a bit further but we can merge what we have here and go ahead in a follow up PR.
I just remember that I opened a PR recently to make vars in this package private. It would have been nice to not introduce a new public var, but we can completely hide `DefaultClient` later, no worries.